pkg/ice/candidate.go

```diff
@@ -0,0 +1,55 @@
+package ice
+
+import (
+	"fmt"
+	"math/rand"
+)
+
+const (
+	hostCandidatePreference  uint16 = 126
+	srflxCandidatePreference uint16 = 100
+)
+
+// Candidate represents an ICE candidate
+type Candidate interface {
+	String(component int) string
+}
+
+// CandidateBase represents an ICE candidate, a base with enough attributes
+// for host candidates, see CandidateSrflx and CandidateRelay for more
+type CandidateBase struct {
+	Protocol TransportType
+	Address  string
+	Port     int
+}
+
+func (c *CandidateBase) priority(typePreference uint16, component uint16) uint16 {
+	localPreference := uint16(rand.Uint32() / 2)
+	return (2^24)*typePreference +
+		(2^8)*localPreference +
+		(2^0)*(256-component)
+}
+
+// CandidateHost is a Candidate of typ Host
+type CandidateHost struct {
+	CandidateBase
+}
+
+// String for CandidateHost
+func (c *CandidateHost) String(component int) string {
+	return fmt.Sprintf("udpcandidate %d udp %d %s %d typ host generation 0",
```

Comment:
I have mixed feelings about this approach. It seems rather clean but this way we are leaking SDP details into the ICE package. E.g.: If we ever want to support the [ORTC API](https://draft.ortc.org/#rtcicecandidate*) this string format doesn't apply anymore.

Reply:
I agree!

Are you ok with the SDP package importing ICE? I can have it case off the Candidate type in there.

Then it can generate the string exactly the same way (and then in the future we can add ORTC support)

Reply:
That does seem more appropriate. ICE does seem like a pretty big dependency for the SDP package but we can worry about that another time.

Reply:
Done!